KAFKA-15995: Initial API + make Producer/Consumer plugins Monitorable

mimaison opened the pull request:
First part of [KIP-877](https://cwiki.apache.org/confluence/display/KAFKA/KIP-877%3A+Mechanism+for+plugins+and+connectors+to+register+metrics). 

This adds the public APIs and enables the following plugins:
- Serializer
- Deserializer
- ProducerInterceptor
- ConsumerInterceptor
- Partitioner
which are all the Producer and Consumer specific plugins.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

mimaison committed:
KAFKA-15995: Initial API + make Producer/Consumer plugins Monitorable

showuon commented on clients/src/main/java/org/apache/kafka/common/internals/Plugin.java:
```diff
@@ -0,0 +1,78 @@
+    @Override
+    public void close() throws Exception {
+        if (pluginMetrics.isPresent()) pluginMetrics.get().close();
+        if (instance instanceof AutoCloseable) ((AutoCloseable) instance).close();
```
In the [KIP](https://cwiki.apache.org/confluence/display/KAFKA/KIP-877%3A+Mechanism+for+plugins+and+connectors+to+register+metrics#KIP877:Mechanismforpluginsandconnectorstoregistermetrics-Supportedplugins), we said:
`The goal is allow this feature to be used by all plugins that are Closeable, AutoCloseable or have a close() methods.`

But here, we only close if it's `Closeable/AutoCloseable` instance. Are we sure all the interfaces having `close()` method is `Closeable/AutoCloseable`?

mimaison replied:
So far the only plugins that support this feature are all AutoCloseable. When we'll add support to plugin that don't implement Closeable/AutoCloseable, we'll need to adjust this method. But I think for now it's fine.

showuon replied:
If someday we missed the "new added plugin without `AutoCloseable`", it might cause the resource leak. I think we should either update the KIP and say "we only support plugins are Closeable, AutoCloseable", or we fix the issue here, with reflection or something else.

showuon commented on clients/src/test/java/org/apache/kafka/common/internals/PluginTest.java:
```diff
@@ -0,0 +1,124 @@
+    private void checkPlugin(Plugin<SomeMonitorablePlugin> plugin, SomeMonitorablePlugin instance, boolean metricsSet) throws Exception {
+        assertSame(instance, plugin.get());
+        if (metricsSet) {
+            assertNotNull(instance.pluginMetrics);
```
I think we should also make sure the content of the `pluginMetrics` is expected. Ex: we can add some sensors, tags and verify they are added in the pluginMetrics.

mimaison replied:
We do that in `PluginMetricsImplTest` and in the test classes for each supported plugin. So I thought focusing this test class on the wrapping and closing of plugin instances made more sense.

showuon replied:
OK. Thanks.

showuon reviewed (commented):
Thanks for the PR. Left some comments.

mimaison commented:
cc @tombentley @C0urante since you also voted on the KIP. Thanks

mimaison commented:
I rebased on trunk to resolve the conflicts.

mimaison commented:
Pinging @gharris1727 as you were also involved in the discussion thread.

gharris1727 commented on clients/src/main/java/org/apache/kafka/common/internals/Plugin.java:
```diff
@@ -0,0 +1,78 @@
+
+    @Override
+    public void close() throws Exception {
+        if (pluginMetrics.isPresent()) pluginMetrics.get().close();
```
Is there a way for the plugin metrics to throw on close? If so, the instance itself wouldn't get closed, leaking it.

I think propagating the exception from the actual instance close is a good idea.

mimaison replied:
Yes I change the logic to capture exceptions thrown by the plugin or the PluginMetrics instance and rethrow the first one.

gharris1727 commented on clients/src/main/java/org/apache/kafka/common/metrics/PluginMetrics.java:
```diff
@@ -0,0 +1,75 @@
+
+/**
+ * This allows plugins to register metrics and sensors.
+ * Any metrics registered by the plugin is automatically removed when the plugin is closed.
```
```suggestion
 * Any metrics registered by the plugin are automatically removed when the plugin is closed.
```

gharris1727 commented on clients/src/main/java/org/apache/kafka/common/internals/Plugin.java:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    @Override
+    public void close() throws Exception {
```
The close order here has some trade-offs.

If we close metrics first, the plugin close() method can't ever emit metrics. I can think of some useful examples that we should permit:
* is a close operation ongoing
* how many records were processed during the instance lifetime
* how long was the plugin open
* how long did operation X take during closing

Also if there are background threads within a plugin, those threads may want to emit metrics concurrent with the close() call. The close() call may leave that background thread running, intentionally or accidentally. I think that once the close() call completes, we can throw away any further metrics from the plugin, making all of the calls no-ops, or maybe throw exceptions if we want to be harsh.

If we close the plugin first, the MetricValueProvider interface may have an undefined value during and after closing. If the plugin needs to be open in order to provide real data, they MetricValueProvider instance may throw an exception, or provide junk or stale data.

I think we should close the plugin first, and mention that the MetricValueProviders may be called at any time, including after the metric is removed or the plugin is closed.

mimaison replied:
That's a good point! And thanks for taking the time to consider the alternatives and suggest solutions.

I agree that the plugin instance should be closed first. This allows the plugin `close()` method to update metrics. Also when closing the PluginMetrics instance afterwards, this allows to "lock" it as in theory it should not receive further calls.

I've opted to use `Utils.closeQuietly()` so we can capture if any of the plugin or PluginMetrics instance fail closing.

gharris1727 commented on clients/src/main/java/org/apache/kafka/common/metrics/internals/PluginMetricsImpl.java:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    @Override
+    public void close() throws IOException {
```
What if additional metrics are added or removed after close() is called?

mimaison replied:
I think I prefer making the contract strict, and disallow any calls once `close()`has been called. So let's throw if this happens.

gharris1727 reviewed (commented):
Thanks for the PR @mimaison!

gharris1727 commented on clients/src/main/java/org/apache/kafka/common/metrics/internals/PluginMetricsImpl.java:
```diff
@@ -0,0 +1,113 @@
+
+    private static final String GROUP = "plugins";
+
+    private final Metrics metrics;
+    private final Map<String, String> tags;
+    private final Set<MetricName> metricNames = new HashSet<>();
+    private final Set<String> sensors = new HashSet<>();
+    private boolean closing = false;
```
I think that this implementation should be thread-safe, at minimum the `closing` variable should be volatile, but maybe the Sets need to be thread-safe as well.

mimaison replied:
Thanks for the review, I pushed an update.

mimaison committed:
Address feedback from Greg

mimaison commented:
@showuon @gharris1727 Thanks for the reviews! Can you take another look?

gharris1727 commented:
@mimaison I don't have any more comments, I think i'm ready to approve this. Can you check the build failures? I've rerun it several times and the ClientUtilsTest seems to be consistently failing, but i'm not sure how your changes could be causing it.

mimaison commented:
I think that ClientUtilsTest issue has been fixed in https://github.com/apache/kafka/pull/18549.
I rebased the PR on trunk and fixed a compilation issue due to a new test in KafkaProducerTest, that should clear the CI issue.

mimaison committed:
Fix compilation issue in KafkaProducerTest

mimaison commented:
We still have a bunch of failures in the CI but these all look unrelated to this PR.
The failure `AbstractCoordinatorTest > testWakeupAfterSyncGroupReceived()` is tracked in https://issues.apache.org/jira/browse/KAFKA-18310.

gharris1727 reviewed (approved):
LGTM.

I was having pretty good luck earlier this month with green builds, so builds seem less reliable than usual. We can merge and monitor trunk, unblocking the next PR.

Thanks @mimaison for the KIP and implementation! So happy to see this change.
